Theory-on-regression.md

```diff
@@ -0,0 +1,97 @@
+
+$$ \color{red} \text{Goal:  } \underset{w}{\text{minimize }} J(w)$$
+
+In this section, we'll learn two regression model — the ordinary least squares for linear regression and neural network for non-linear regression — as well as related loss function and optimization algorithms.
+
+
+## Ordinary Least Squares (linear regression)
+
+Linear here means that the target value is expected to be a linear combination of the input features.
+
+### Model representation
+
+The model of linear regression is
+$$\hat{y}(w,x) = w_0 + w_1x_1 + w_2x_2 + \cdots + w_nx_n \$$
+where $\hat{y}$ is the target value, vector $X = (x_1, \cdots, x_n)$ are input features and $W = (w_0, \cdots, w_n)$ are weights. weight $w_0$ is also called intercept, and the other weights $(w_1, \cdots, w_n)$ are coefficients.
+
+If there is only one input feature, the model becomes
+$$\hat{y}(w,x) = w_0 + w_1x_1$$
+it's also called univariate linear regression.
+
+### Loss function
+
+Loss function is to measure the "error" between real values and predicted values. For regression, it can be presented as the following form:
+
+$$ J(w) = ||y - \hat{y}||^2 $$
+
+which is the sum of squares of the differences between real values $y$ and predicted values $\hat{y}$.
+
+### Optimization algorithm
+
+Optimization algorithm is the algorithms used to minimize the loss function, i.e. to keep changing weights $w$ to reduce loss $J(w)$ until it hopefully ends up at a minimum.
+
+The most basic and popular optimization algorithm is [gradient descent](https://en.wikipedia.org/wiki/Gradient_descent), also known as steepest descent. It can be described as below:
+
+$$ \color{gray} \text{repeat until convergence: } \\
+     \color{black} w_i := w_i - \alpha\frac{\partial{}}{\partial{w}}J(w) \\
+    \color{gray}
+    \text{for } i = (0, \cdots, n) \\
+    \alpha \text{ is learning rate}
+$$
+
+The basic idea is to take repeated steps in the opposite direction of the gradient of the loss function, i.e. $\displaystyle -\frac{\partial{}}{\partial{w}}J(w)$, which will lead to a local minimum of loss function, as shown below.
+
+![Gradient descent 1D](image/Gradient-descent.png)
+<!-- Figure Gradient-descent.png from https://imaddabbura.github.io/img/gradient-descent-algorithms/gradients.PNG -->
+
+The process of gradient descent for two features would look like the dynamic diagram below:
+![Gradient descent 2D dynamic](image/Gradient_descent_dynamic.gif)
+<!-- Figure Gradient_descent_dynamic.gif from https://blog.paperspace.com/intro-to-optimization-in-deep-learning-gradient-descent/ -->
+
+
+## Neural network (non-linear regression)
+
+### Model representation
+
```

Comment:
Here we can borrow from the Deep Learning lesson, see [this part](https://github.com/carpentries-incubator/deep-learning-intro/blob/gh-pages/_episodes/01-introduction.md#neural-networks). I think we can copy most of it, I find the illustration of a single neuron also very useful.